### PR TITLE
Kernel/HID: Build better abstractions in the subsystem, stop use LockRefPtrs

### PR DESCRIPTION
**Pull request by supercomputer7**
TL;DR - This is somewhat a preparation for a work I do for VirtIO HID PCI devices, but in general is also a valid effort through removing LockRefPtrs too.

----

For a very long time, the kernel had only support for basic PS/2 devices
such as the PS2 AT keyboard and regular PS2 mouse (with a scroll wheel).

To adapt to this, we had very simple abstractions in place, essentially,
the PS2 devices were registered as IRQ handlers (IRQ 1 and 12), and when
an interrupt was triggered, we simply had to tell the I8042Controller to
fetch a byte for us, then send it back to the appropriate device for
further processing and queueing of either a key event, or a mouse packet
so userspace can do something meaningful about it.

When we added the VMWare mouse integration feature it was easily adapted
to this paradigm, requiring small changes across the handling code for
these devices.

This patch is a major cleanup for any future advancements in the HID
subsystem.
It ensures we do things in a much more sane manner:
- We stop using LockRefPtrs. Currently, after the initialization of the
  i8042 controller, we never have to change RefPtrs in that class, as we
  simply don't support PS2 hotplugging currently.
- There's a clear separation between PS2 devices and the actual device
  nodes that normally exist in /dev. PS2 devices are not polled, because
  when the user uses these devices, they will trigger an IRQ which when
  is handled, could produce either a MousePacket or KeyEvent, depending
  on the device state.
  The separation is crucial for buses that are polled, for example - USB
  is a polled bus and will not generate an IRQ for HID devices.
- There's a clear separation in roles of each structure. The PS2 devices
  which are attached to a I8042Controller object are managing the device
  state, while the generic MouseDevice and KeyboardDevice manage all
  related tasks of a CharacterDevice, as well as interpreting scan code
  events.

P.S. The separation in roles of each structure paves the path for adding support for HID devices that are attached to busses like VirtIO, USB, etc. It also opens an opportunity to do automated input tests for HID devices much more easily.
We will see how things rolls in the long-term, but for now we should keep this PR in this form :)

### CODE REVIEW

**Comment by supercomputer7** [comment]
I might keep the last commit for a future PR if reviewers are not convinced that we should add it right now.

**Comment by supercomputer7** [comment]
> I might keep the last commit for a future PR if reviewers are not convinced that we should add it right now.

On second thought, I could make it even better. This will be a draft but hopefully soon it will be undrafted and more LockRefPtrs will be gone :)

File: Kernel/Arch/x86_64/ISABus/I8042Controller.cpp [hunk 1]
```diff
@@ -19,15 +19,6 @@ UNMAP_AFTER_INIT NonnullLockRefPtr<I8042Controller> I8042Controller::initialize(
...
 
-LockRefPtr<MouseDevice> I8042Controller::mouse() const
-{
-    return m_mouse_device;
-}
-LockRefPtr<KeyboardDevice> I8042Controller::keyboard() const
-{
-    return m_keyboard_device;
-}
-
```
Comment (GMTA) [inline Kernel/Arch/x86_64/ISABus/I8042Controller.cpp:30]: You should remove these methods in the same commit where you stop using them.

File: Kernel/Arch/x86_64/ISABus/HID/PS2KeyboardDevice.cpp [hunk 2]
```diff
@@ -23,89 +21,46 @@ void PS2KeyboardDevice::irq_handle_byte_read(u8 byte)
 {
     u8 ch = byte & 0x7f;
     bool pressed = !(byte & 0x80);
-
-    m_entropy_source.add_random_event(byte);
-
```
Comment (GMTA) [inline Kernel/Arch/x86_64/ISABus/HID/PS2KeyboardDevice.cpp:28]: You remove these `add_random_event` invocations without mention in the commit message - it seems useful to take user input into account for entropy sources, so we either need to replace this with something equivalent or explain why we're removing this.

Comment (supercomputer7) [reply Kernel/Arch/x86_64/ISABus/HID/PS2KeyboardDevice.cpp:28]: Oh I forgot about those! I will hook it in the generic `KeyboardDevice` and `MouseDevice` classes :)

File: Kernel/Arch/x86_64/ISABus/I8042Controller.cpp (hunk 1 above)
Comment (supercomputer7) [reply Kernel/Arch/x86_64/ISABus/I8042Controller.cpp:30]: Done.

File: Kernel/Arch/x86_64/ISABus/HID/PS2KeyboardDevice.cpp (hunk 2 above)
Comment (supercomputer7) [reply Kernel/Arch/x86_64/ISABus/HID/PS2KeyboardDevice.cpp:28]: I hooked this into the generic HID devices, which should be enough :)

File: Kernel/Arch/x86_64/ISABus/I8042Controller.cpp (hunk 1 above)
Comment (GMTA) [reply Kernel/Arch/x86_64/ISABus/I8042Controller.cpp:30]: This is not the right commit - you removed usage of them in 35bfa4a7538686431db19da03ecc1e17291c28f4

**Review by GMTA — APPROVED** [review]
